### ISSUE TITLE
Refresh token doesn't work

### ISSUE THREAD

**Issue by niiamon** [doorkeeper-gem/doorkeeper#739]
For some reason, refresh tokens have stopped working for me. Every request to refresh a token results in a `{"error":"invalid_grant","error_description":"The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."}`.

I think my setup is correct:
- `use_refresh_token` inside my doorkeeper.rb initializer file
- sample refresh toke n request:
  `curl -v -H 'Content-Type: application/json' -H 'Accept: application/json' -X POST "http://localhost:3002/oauth/token" -d "{\"grant_type\": \"refresh_token\", \"client_id\": \"fa927b2631bed16e3766318cc3db4910f003e9c9fc8a57eaaf2e9490a9436615\", \"client_secret\": \"b1d402d325ed2fb58ac9342ade12cdf1664716d9df4327f092fffbf5c08259b4\", \"refresh_token\": \"4b9d40942ba6d47c30c808253c18416085bd4b724ddebd741114ac7ce74690fd\"}"`
- the token matching this refresh token exists within `oauth_access_tokens` table

I have read that refresh tokens do not expire, as can be expected. However, apart from an expired refresh token, I really cannot see what the issue is. I have tried adding the `redirect_uri` to the request with no change in outcome. What is the recommended way to getting refresh tokens to work?

**Comment by mindeavor** [comment]
[According to the docs](https://github.com/doorkeeper-gem/doorkeeper/wiki/Enable-Refresh-Token-Credentials#doorkeeper-implementation), doorkeeper expires a refresh token when you use it to get a new access token.

**Comment by niiamon** [comment]
That makes a lot of sense. But I don't recall using that refresh token to get a new access token. When I try to issue a new access token using that unused refresh token, I always get that response.

**Comment by d4x337** [comment]
Refresh Token is not working also here. After requesting it i got a 401. Here is the server log output:

Started POST "/oauth/token" for 127.0.0.1 at 2015-11-05 10:45:22 +0100
Processing by Doorkeeper::TokensController#create as */*
  Parameters: {"grant_type"=>"refresh_token", "client_id"=>"abc38d2ec3362329294f4340a62e19513ab311f9a28d07b2eb9ae796507aa695", "client_secret"=>"[FILTERED]", "refresh_token"=>"[FILTERED]"}
  Doorkeeper::AccessToken Load (0.8ms)  SELECT  "oauth_access_tokens".* FROM "oauth_access_tokens"  WHERE "oauth_access_tokens"."refresh_t
oken" = 'f97841e04014f576bdc5f56aa96c98ea87f8cae61c137cb56d85d1358c1ee722'  ORDER BY "oauth_access_tokens"."id" ASC LIMIT 1
  Doorkeeper::Application Load (0.6ms)  SELECT  "oauth_applications".* FROM "oauth_applications"  WHERE "oauth_applications"."uid" = 'abc3
8d2ec3362329294f4340a62e19513ab311f9a28d07b2eb9ae796507aa695' AND "oauth_applications"."secret" = '428ef5c6879ceb0fc279b739934957dc7d9ff6c
c85197375d3a020c3f4a55863' LIMIT 1
Completed 401 Unauthorized in 256ms

**Comment by d4x337** [comment]
Hi guys, 
I have solved by adding in the curl request the client_id and secret for a password_grant.
This way the application_id in oauth_token_applications is properly inserted and everything work fine.

**Comment by tute** [comment]
Is this a bug in doorkeeper, or do you need help with your configuration or setup? Thanks!

**Comment by d4x337** [comment]
No I have updated the ticket. If there is a bug it is in the doc, grant for
refresh token needs client id and secret to work, not like in the example.

Have a nice day

Davide

Il lunedì 23 novembre 2015, Tute Costa <notifications@github.com> ha
scritto:

**Comment by niiamon** [comment]
@d4x337 awesome! I'm going to give this a shot. I suppose it's not strange that the `client_id` and `client_secret` are required when you attempt a token refresh.

**Comment by niiamon** [comment]
@d4x337 could I see a sample URL you used? To confirm on my end.

**Comment by tute** [comment]
This issue seems the contrary to https://github.com/doorkeeper-gem/doorkeeper/issues/874. What is the case? Do we require or not client authentication? Remember that this will depend on the type of grant being used. Thanks!

**Comment by eebasadre** [comment]
@niiamon I have the same problem to you know and I wonder how you manage to solve the issue which is before refresh token is working properly then now it stop working. Can I have your request parameters?

**Comment by eebasadre** [comment]
@d4x337 sorry, I didn't get you why you change the grant type from refresh_token to password? Can I see a sample request parameters you used?

**Comment by eebasadre** [comment]
Hi, I managed to fix it just now. 
This is my params that I passed.

## Parameter: 

```
{"refresh_token"=>"b9a9525d916696d5028a39bc2533...",  
"format"=>:json,  
"controller"=>"/oauth/tokens",  
"action"=>"create",  
"grant_type"=>"refresh_token"
}
```
Then in my headers I set the `grant_type: refresh_token`, `Accept: application/json`, `Authorization Bearer {{user_token}}` the expired user token and in the body the refresh_token: `refresh_token`. So, I don't need to pass the client id and secret. However, I'm still confused why @niiamon passed the client id and secret then changed the grant type to password.

**Comment by nbulaj** [comment]
Hi @niiamon . Did you solve the problem?

**Comment by mantrala** [comment]
I just ran in to this issue and solved it. For anyone making their way here:

The docs say the required params are only: `grant_type`, `client_id`, `client_secret` and `refresh_token`.

It turned out that we had a custom `Oauth::TokensController` which was overriding `create` method and was not handling the `grant_type` of `refresh_token`.

Changed the controller to inherit from `Doorkeeper::TokensController` and calling `super` for whatever `grant_type` I wanted to handle. Also, make sure that `use_refresh_token` is enabled in the doorkeeper initializer.

Example:

`class TokensController < Doorkeeper::TokensController`

example request:
```
curl -X POST \
  http://localhost:3000/oauth/token \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'grant_type=refresh_token&client_id=my_client_id&client_secret=my_secret_id&refresh_token=my_refresh_token'
```
